whylog/constraints/__init__.py

```diff
@@ -0,0 +1,36 @@
+class Constraint(object):
```

Comment:
constraint object could be added to rule object - this would be a method of Rule class in such case

AbstractConstraint must define some kind of interface for checking if it is, at least for testing during teaching. Maybe `.is_satisfied() -> bool`?